File: app/src/main/java/com/akari/ppx/xp/hook/code/assist/SendGodHook.java
package com.akari.ppx.xp.hook.code.assist;

import com.akari.ppx.common.utils.XSP;
import com.akari.ppx.xp.hook.code.SuperbHook;

import de.robv.android.xposed.XC_MethodHook;

import static com.akari.ppx.common.constant.Prefs.AUTO_SEND_GOD_ENABLE;
import static com.akari.ppx.common.constant.Prefs.AUTO_SEND_GOD_TIME_LIMIT;
import static com.akari.ppx.common.constant.Prefs.UNLOCK_SEND_GOD;
import static de.robv.android.xposed.XposedHelpers.callMethod;
import static de.robv.android.xposed.XposedHelpers.findClass;
import static de.robv.android.xposed.XposedHelpers.getBooleanField;
import static de.robv.android.xposed.XposedHelpers.getIntField;
import static de.robv.android.xposed.XposedHelpers.getLongField;
import static de.robv.android.xposed.XposedHelpers.newInstance;
import static de.robv.android.xposed.XposedHelpers.setLongField;

public class SendGodHook extends SuperbHook {
	@Override
	protected void onHook(ClassLoader cl) {
		final boolean autoSendGod = XSP.get(AUTO_SEND_GOD_ENABLE);
		final int timeLimit = XSP.getI(AUTO_SEND_GOD_TIME_LIMIT, 21600);
		hookMethod(UNLOCK_SEND_GOD, "com.sup.android.mi.feed.repo.bean.comment.Comment", "getSendGodStatus", new XC_MethodHook() {
			@Override
			protected void afterHookedMethod(MethodHookParam param) {
				if (getLongField(param.thisObject, "aliasItemId") == 1 && !getBooleanField(param.thisObject, "hasLiked")) {
					setLongField(param.thisObject, "aliasItemId", 2);
					callMethod(newInstance(findClass("com.sup.android.detail.util.o", cl)), "a",
							8, getLongField(param.thisObject, "commentId"), true, 10, 1);
					if (autoSendGod && System.currentTimeMillis() / 1000 - getLongField(param.thisObject, "createTime") <= timeLimit)
						callMethod(newInstance(findClass("com.sup.android.m_comment.util.helper.d", cl)), "a",
								getLongField(param.thisObject, "itemId"), getLongField(param.thisObject, "commentId"), null);
					param.setResult(1);
					return;
				}
				if (getIntField(param.thisObject, "sendGodStatus") == 3) {
					if (getLongField(param.thisObject, "aliasItemId") == 2) {
						param.setResult(autoSendGod && System.currentTimeMillis() / 1000 - getLongField(param.thisObject, "createTime") <= timeLimit ? 2 : 1);
						return;
					}
					setLongField(param.thisObject, "aliasItemId", 1);
					param.setResult(1);
				}
			}
		});

	}
}

File: app/src/main/java/com/akari/ppx/xp/hook/assist/SendGodHook.kt
@file:Suppress("unused")

package com.akari.ppx.xp.hook.assist

import com.akari.ppx.data.XPrefs
import com.akari.ppx.utils.*
import com.akari.ppx.xp.Init.cl
import com.akari.ppx.xp.hook.SwitchHook

class SendGodHook : SwitchHook("unlock_send_god_limit") {
    override fun onHook() {
        val isAuto = XPrefs<Boolean>("auto_send_god")
        val autoTimeLimit = XPrefs<String>("auto_send_god_time_limit").let {
            if (it.isEmpty()) 86400 else it.toInt()
        }

        fun Any.shouldSendGod() =
            isAuto && System.currentTimeMillis() / 1000 - getLongField("createTime") <= autoTimeLimit

        "com.sup.android.mi.feed.repo.bean.comment.Comment".hookAfterMethod(
            cl,
            "getSendGodStatus"
        ) { param ->
            with(param.thisObject) {
                checkIf({ getLongField("aliasItemId") == 1L && !getBooleanField("hasLiked") }) {
                    setLongField("aliasItemId", 2)
                    diggCell(cellId = getLongField("commentId"))
                    if (shouldSendGod())
                        diggGodComment(getLongField("itemId"), getLongField("commentId"))
                    param.result = 1
                    return@hookAfterMethod
                }
                checkIf({ getIntField("sendGodStatus") == 3 }) {
                    if (getLongField("aliasItemId") == 2L) {
                        param.result = if (shouldSendGod()) 2 else 1
                        return@hookAfterMethod
                    }
                    setLongField("aliasItemId", 1)
                    param.result = 1
                }
            }

        }
    }
}